app/main.py

```diff
@@ -1,34 +1,117 @@
+        decks = []
+        if self.start[0] == self.end[0]:
+            for column in range(self.start[1], self.end[1] + 1):
+                decks.append(Deck(self.start[0], column))
+        elif self.start[1] == self.end[1]:
+            for row in range(self.start[0], self.end[0] + 1):
+                decks.append(Deck(row, self.start[1]))
+        return decks
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def get_deck(self, row: int, column: int) -> Deck | None:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+        return None
+
+    def fire(self, row: int, column: int) -> bool:
+        deck = self.get_deck(row, column)
+        if deck and deck.is_alive:
+            deck.is_alive = False
+            if all(not d.is_alive for d in self.decks):
+                self.is_drowned = True
+            return True
+        return False
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list[tuple[tuple[int, int]]]) -> None:
+        self.size = 10
+        self.ship_cells = [
+            ["~" for _ in range(self.size)] for _ in range(self.size)
+        ]
+        if all(
+                isinstance(cords, tuple) and len(cords) == 2
+                for ship in ships for cords in ship
+        ):
+            self.ships = [
+                Ship(start, end) for start, end in ships
+            ]
+        self.field = {}
+        self._place_ships()
+        self._validate_field()
+
+    def fire(self, location: tuple) -> str:
+        row, column = location
+        if (row, column) not in self.field:
+            return "Miss!"
+        ship = self.field[(row, column)]
+        if ship.fire(row, column):
+            if ship.is_drowned:
+                for deck in ship.decks:
+                    self.ship_cells[deck.row][deck.column] = "x"
+                return "Sunk!"
+            self.ship_cells[row][column] = "*"
+            return "Hit!"
+        return "Miss!"
+
+    def _place_ships(self) -> None:
+        for ship in self.ships:
+            for deck in ship.decks:
+                self.ship_cells[deck.row][deck.column] = u"\u25A1"
+                self.field[(deck.row, deck.column)] = ship
+
+    def _validate_field(self) -> None:
+        if len(self.ships) != 10:
+            raise ValueError("Total number of ships must be 10")
```

Comment:
The `_validate_field` method assumes there are exactly 10 ships. Ensure that the input list of ships adheres to this requirement, or adjust the validation logic to match the intended game rules.